cmd/run.go

```diff
@@ -37,12 +41,36 @@ type cmdRun struct {
 //
 //nolint:funlen,gocognit,gocyclo,cyclop
 func (c *cmdRun) run(cmd *cobra.Command, args []string) (err error) {
+	var logger logrus.FieldLogger = c.gs.logger
+	defer func() {
+		logger.Debugf("Everything has finished, exiting k6 with error '%s'!", err)
+	}()
 	printBanner(c.gs)
 
+	globalCtx, globalCancel := context.WithCancel(c.gs.ctx)
+	defer globalCancel()
+
+	// lingerCtx is cancelled by Ctrl+C, and is used to wait for that event when
+	// k6 was started with the --linger option.
+	lingerCtx, lingerCancel := context.WithCancel(globalCtx)
+	defer lingerCancel()
+
+	// runCtx is used for the test run execution and is created with the special
+	// execution.NewTestRunContext() function so that it can be aborted even
+	// from sub-contexts while also attaching a reason for the abort.
+	runCtx, runAbort := execution.NewTestRunContext(lingerCtx, logger)
```

Comment:
I think we are not cancelling the `runCtx` so the Report's goroutine is stuck waiting for it.

Reply:
hmm I think the bug is that it actually tries to wait for `<-runCtx.Done()` at all :thinking: 

https://github.com/grafana/k6/blob/3c00d584bd1e87c8350d75fc168cfdbe6f9a3104/cmd/run.go#L285-L302

We call it after the test run has already finished, so there is no point in waiting for `<-runCtx.Done()`, we should just send the report :thinking: This bug is just a remnant from the old way convoluted way...

And we can wait for it to be done with a separate `defer`, "earlier" in the process. That is, k6 can continue crunching metrics, showing the summary, stopping outputs, etc., while the report is being sent.

Reply:
I think this should now be fixed by the [latest force-push](https://github.com/grafana/k6/compare/3c00d584bd1e87c8350d75fc168cfdbe6f9a3104..6f6b105ff041b2c317fa909d959bb77ea0946773)?

Reply:
It is yes.